### PR TITLE
ENH: QUAST outputs downloaded genome artifacts

### LINKED ISSUE #93: ENH: output fetched reference genomes from 'evaluate-quast'
**Is your feature request related to a problem? Please describe.**
Whenever QUAST is run and reference genomes are not provided it will try to identify some references and fetch them from NCBI. Since all of those are saved to a temporary directory, they will be removed after the action completes.

**Describe the solution you'd like**
It would be great if we could collect those genomes and output them as `GenomeData[DNASequence]` artifact - that way, next time when a user wants to re-run QUAST for whatever reason, they can just pass those references instead of re-downloading all of them.

**Additional context**
The references fetched by QUAST are combined into a single file which can be found in `results/quast_downloaded_references` (?). Alternatively, individual references could also be passed - not sure what kind of artifact they should be stored in, though.

### PR DESCRIPTION
**Pull request by DorielaGrabocka**
Closes #93

### CODE REVIEW

File: q2_assembly/plugin_setup.py [hunk 1]
```diff
@@ -179,11 +180,15 @@
...
-        "references": List[FeatureData[Sequence]],
+        "references": GenomeData[DNASequence],
         "mapped_reads": SampleData[AlignmentMap],
     },
     parameters=quast_params,
-    outputs={"results_table": QUASTResults, "visualization": Visualization},
+    outputs=[
+        ("results_table", QUASTResults),
+        ("visualization", Visualization),
+        ("genomes", GenomeData[DNASequence]),
```
Comment (misialq) [inline q2_assembly/plugin_setup.py:190]: Maybe we can call this output `reference_genomes` just to be a bit more explicit.

File: q2_assembly/quast/quast.py [hunk 2]
```diff
@@ -227,6 +236,14 @@ def _zip_additional_reports(path_to_dirs: list, output_filename: str) -> None:
             _zip_dir(zipf, directory)
 
 
+def _move_references(src, dest):
+    file_patterns = ["*.fasta", "*.fa"]
+    os.makedirs(os.path.join(dest, "quast_downloaded_references"), exist_ok=True)
+    for pattern in file_patterns:
+        for fp in glob.glob(os.path.join(src, pattern)):
+            shutil.copy(fp, os.path.join(dest, "quast_downloaded_references"))
```
Comment (misialq) [inline q2_assembly/quast/quast.py:244]: This loop is a bit redundant, given the two patterns you have here: if you have a `*.fa` pattern - it should already contain all the `*.fasta` files so there should be no need to loop through both.

File: q2_assembly/quast/quast.py [hunk 3]
```diff
@@ -159,13 +161,20 @@ def _evaluate_contigs(
     if references:
         all_refs_dir = os.path.join(results_dir, "references")
         os.makedirs(all_refs_dir, exist_ok=True)
+        copy_tree(str(references), all_refs_dir)
```
Comment (misialq) [inline q2_assembly/quast/quast.py:164]: Why do you need to copy this? This was required when we had files containing multiple genomes per fasta file which had to be split into individual files. Your genomes already are separated so you shouldn't need to copy anything - you just need to gather paths to those and pass them to quast below.

File: q2_assembly/quast/quast.py [hunk 4]
```diff
@@ -295,6 +312,13 @@ def _visualize_quast(
         # Copy results to output dir
         copy_tree(results_dir, os.path.join(output_dir, "quast_data"))
 
+        # Save the downloaded references
+        if not references:
+            download_references = os.path.join(
+                results_dir, "quast_downloaded_references"
+            )
+            _move_references(download_references, output_dir)
+
```
Comment (misialq) [inline q2_assembly/quast/quast.py:321]: What would happen here if there was no internet connection _and_ the user did not provide references?

**Review by misialq — CHANGES_REQUESTED** [review]
Hey @DorielaGrabocka, thanks for the PR! See some change suggestions below. One question, though: looking through this made me realize that if we do it this way the genomes will become part of the actual visualization artifact, right? This is something that absolutely should not happen, so we need to find another way...

**Comment by DorielaGrabocka** [comment]
BLASTN DB issue pending.

File: q2_assembly/quast/quast.py [hunk 5]
```diff
@@ -383,4 +409,17 @@ def evaluate_contigs(
         # 3. read it as a pandas dataframe then we create the QUASTResults
         tabular_results = ctx.make_artifact("QUASTResults", report_df)
 
-    return tabular_results, visualization
+        # 4. create the Genomes
+        if not references:
+            try:
+                genomes = ctx.make_artifact("GenomeData[DNASequence]", genomes_dir)
+            except ValidationError:
```
[the remark this answers is not in the document]
Comment (misialq) [reply q2_assembly/quast/quast.py:416]: Also, you don't need to rely on the validation to detect the first case - you can first check whether the `genomes_dir` is empty, no?

File: q2_assembly/quast/quast.py [hunk 6]
```diff
@@ -244,14 +247,23 @@ def _visualize_quast(
     reads: Union[
         SingleLanePerSamplePairedEndFastqDirFmt, SingleLanePerSampleSingleEndFastqDirFmt
     ] = None,
-    references: DNAFASTAFormat = None,
+    references: GenomeSequencesDirectoryFormat = None,
+    genomes_out_path: str = None,
```
Comment (misialq) [inline q2_assembly/quast/quast.py:251]: Maybe we can shorten this to something like `genomes_dir`?

File: q2_assembly/quast/quast.py [hunk 7]
```diff
@@ -382,4 +412,34 @@ def evaluate_contigs(
...
+        if not references:
+            try:
+                genomes = ctx.make_artifact("GenomeData[DNASequence]", genomes_dir)
+            except ValidationError as e:  # corrupt files
+                if "Missing one or more" in str(e):
+                    warn(
+                        "WARNING: QUAST did not download any genomes. The returned "
+                        "GenomeData[DNASequence] artifact is empty. Please check "
+                        "the network connection or provide the reference genomes."
+                    )
```
Comment (misialq) [inline q2_assembly/quast/quast.py:425]:
```suggestion
                        "QUAST did not download any genomes, most likely due to a network "
                        "connection problem. The returned GenomeData[DNASequence] artifact "
                        "will be empty. Please check the network connection or provide the "
                        f"reference genomes. The original error was: '{e}'."
                    )
```

File: q2_assembly/quast/quast.py [hunk 8]
```diff
@@ -382,4 +412,34 @@ def evaluate_contigs(
...
+                else:
+                    warn(
+                        "WARNING: There is a problem with the downloaded genome files. "
+                        "The returned GenomeData[DNASequence] artifact will be empty. "
+                        "You can either try again or provide the reference genomes."
+                    )
```
Comment (misialq) [inline q2_assembly/quast/quast.py:431]: I think this last line is a bit confusing. You should rather say something like:
```suggestion
                    warn(
                        "There was a problem with the genome files downloaded by QUAST. "
                        "The returned GenomeData[DNASequence] artifact will be empty. "
                        f"The original error was: '{e}'."
                    )
```

File: q2_assembly/quast/quast.py [hunk 9]
```diff
@@ -382,4 +412,34 @@ def evaluate_contigs(
...
+                    # we also need to remove whatever is in the directory that
+                    # will not comply with the GenomeSequencesDirectoryFormat
+                    for f in os.listdir(genomes_dir.path):
+                        fp = os.path.join(genomes_dir.path, f)
+                        if os.path.isdir(fp):
+                            shutil.rmtree(fp)
+                        else:
+                            os.remove(fp)
```
Comment (misialq) [inline q2_assembly/quast/quast.py:439]: Honestly, instead of doing all this, we can also just re-create the genomes_dir (the old one will be cleaned up when the action exits):
```
genomes_dir = GenomeSequencesDirectoryFormat()
```

File: q2_assembly/quast/tests/test_quast.py [hunk 10]
```diff
@@ -12,13 +12,18 @@
...
 import tempfile
 import unittest
+import warnings
 from filecmp import dircmp
 from subprocess import CalledProcessError
 from unittest.mock import ANY, MagicMock, Mock, call, mock_open, patch
 from zipfile import ZipFile
 
 import pandas as pd
+from future.moves import itertools
```
Comment (misialq) [inline q2_assembly/quast/tests/test_quast.py:22]: Why are you using itertools from `future.moves`?

File: q2_assembly/quast/tests/test_quast.py [hunk 11]
```diff
@@ -12,13 +12,18 @@
...
+from globus_sdk.experimental.auth_requirements_error import ValidationError
```
Comment (misialq) [inline q2_assembly/quast/tests/test_quast.py:23]: And why on earth would we want to use this validation error? There is a QIIME-built-in one you should be using (please pay attention to PyCharm autosuggested imports).

**Review by misialq — CHANGES_REQUESTED** [review]

File: q2_assembly/quast/tests/test_quast.py (hunk 10 above)
Comment (DorielaGrabocka) [reply q2_assembly/quast/tests/test_quast.py:22]: Actually this was the only option that offered me the combination of parameters I wanted. I have simplified the parameterization in the new commit and I am removing the import.

Comment (misialq) [reply q2_assembly/quast/tests/test_quast.py:22]: The point is not to remove it. My question is why is the import from `future.moves`? Was it PyCharm's suggestion or ChatGPT's? You should use `product` from `itertools` directly - future.moves is a library that was introduced long time ago for compatibility between Python 2 and 3 and really has no place in our Python 3 code anymore. Or did you try the new method and it didn't work?

Comment (DorielaGrabocka) [reply q2_assembly/quast/tests/test_quast.py]: It was Pycharm's suggestion. It gave me two options and this was the first one. I tried the `product` one before committing and it did not have the `repeat` argument. Actually, for the 3 tuples that I needed for that cases, I believe it was a bit overly complicated😅, that's why I manually wrote them. You can check and let me know what you think.

Comment (misialq) [reply q2_assembly/quast/tests/test_quast.py]: But of course there is a `repeat` argument there - it's the same function... https://docs.python.org/3/library/itertools.html#itertools.product

You're right, though - manual seems a bit simpler 😅

Comment (DorielaGrabocka) [reply q2_assembly/quast/tests/test_quast.py]: Ah never mind, my bad, the second trial was from itertools from `tqdm`, somehow Pycharm does not suggest me the simple built-in itertools...🤔

So I am leaving it manual (it seems cleaner as well😄)

File: q2_assembly/plugin_setup.py [hunk 12]
```diff
@@ -156,21 +157,29 @@
...
         "mapped_reads": "Reads-to-contigs alignment maps (alternative to 'reads')."
         "directly.",
     },
-    parameter_descriptions=quast_param_descriptions,
+    parameter_descriptions={
+        **quast_param_descriptions,
+        "genomes_dir": "Path of the directory from which GenomeData[DNASequence] "
+        "will be created."
+        "artifact that will be returned if"
+        "reference genomes are downloaded.",
```
Comment (misialq) [inline q2_assembly/plugin_setup.py:176]: Hmmm, isn't something wrong with this comment? It looks like there's some text leftover 😅

Comment (DorielaGrabocka) [reply q2_assembly/plugin_setup.py:176]: My bad, sorry!

File: q2_assembly/quast/tests/test_quast.py [hunk 13]
```diff
@@ -757,7 +753,9 @@ def test_create_tabular_results_in_combined_subdir(self, p1, p2):
         p2.assert_called_once_with(report_path, sep="\t", header=0)
         p1.assert_called_once_with(mock_df, [1000, 5000, 25000, 50000])
 
-    def test_evaluate_contigs_pipeline(self):
+    @parameterized.expand([(True, True), (False, True), (False, False)])
+    @patch("warnings.warn")
```
Comment (misialq) [inline q2_assembly/quast/tests/test_quast.py:757]: I must say, that only now that you removed this itertools import I see what's happening in this test. The bad news is - all these cases you are testing here should be _separate_ test cases. This way it is very hard to understand what is actually happening in your test with all these nested if statements. Unit tests should be only as complex as really necessary - please split this into three different, smaller tests (maybe there's something else you can do, though, to avoid code repetition).

File: q2_assembly/quast/tests/test_quast.py [hunk 14]
```diff
@@ -791,15 +794,246 @@ def _copy(dst_dir):
...
+    def test_evaluate_contigs_pipeline_no_refs_normal(self):
+        # this is used for mocking
+        def _copy(dst_dir):
+            os.makedirs(dst_dir, exist_ok=True)
+            shutil.copy2(
+                src=os.path.join(
+                    self.get_data_path("quast-results"), "enhanced_tabular_results.tsv"
+                ),
+                dst=os.path.join(dst_dir, "quast_results.tsv"),
+            )
```
Comment (misialq) [inline q2_assembly/quast/tests/test_quast.py:821]: You can put this method on `self` directly - then it will become accessible in other tests, like below.

File: q2_assembly/quast/tests/test_quast.py [hunk 15]
```diff
@@ -791,15 +794,246 @@ def _copy(dst_dir):
...
+            ctx = Context()
+            ctx._scope = MagicMock()
+            make_artifact = MagicMock(
+                side_effect=lambda *args: ctx.make_artifact(*args)
+            )
+            mock_ctx = MagicMock(
+                spec=Context,
+                get_action=MagicMock(return_value=action),
+                make_artifact=make_artifact,
+            )
```
Comment (misialq) [inline q2_assembly/quast/tests/test_quast.py:844]: You could also write a small method to generate the mocked context and use it in other tests.

**Review by misialq — APPROVED** [review]
Great, thanks @DorielaGrabocka - time to merge! 🚀
